ls1intum/Artemis
# `General`: Prevent invalid date values in edit dialogs

### Checklist
#### General



#### Client

### Motivation and Context


Closes https://github.com/ls1intum/Artemis/issues/6190

### Description

Before it was possible to input arbitrary strings into the date input field and then save an exercise, course, lecture or exam. If the input could not be converted, the corresponding value was set to null. Now in case of an invalid input (that cannot be converted into a date), the input field is marked red, an error appears and the save button is disabled. Any strings that can be converted into a date are still allowed, meaning many date different formats are supported. In case the inputed string is a valid date, it gets converted into the ''MMM D, YYYY, H:mm:ss A' format (e.g Apr 14, 2023, 10:29:00 AM). If the string is invalid, it stays the same.


### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course

1. Log in to Artemis
2. Navigate to Course Management, to the Course
3. Go to an edit dialogue of exercises (excluding programming), courses, lectures and exams
5. Try out different invalid strings, check if the error appears. Also check for the translation of the error message.
6. Check that the save button is disabled
7. Make sure that valid dates get saved correctly
8. Also check that the calendar opens at the date that is currently selected. If no date is selected or if the value in the input field is invalid, the calander should open at the current date and time.

### Review Progress






#### Code Review
#### Manual Tests

### Screenshots


#### Before:
![image](https://user-images.githubusercontent.com/73339358/232021119-b30ef656-ecc6-40e6-9cd0-6e18c7c03593.png)
You can see which values are not accepted as a date by the color red. It is however difficult to know how the other two values are going to be converted into the format used by the calendar. To see this you have to save and exit the exercise.

https://user-images.githubusercontent.com/73339358/232026233-53874766-927e-4d2e-91cb-465311ce1490.mp4




#### Now:
![image](https://user-images.githubusercontent.com/73339358/232020996-572bf04b-5f18-4f38-af6b-ec9c98f63f11.png)
After exiting the input field, the dates automatically get converted into the same format. The user knows directly which date is going to be saved. There are some edge cases, that I was not able to get rid of, e.g the string Nov 31 gets converted into Dec 1. The conversion is done automatically by the framework and I was not able to find a better solution than to just display the date after it was converted.

https://user-images.githubusercontent.com/73339358/232026264-8bf608bc-60ed-4a73-8c30-ac9ebdf31250.mp4


### Test Coverage
| Class/File | Line Coverage | Confirmation (assert/expect) | Comment |
|------------|--------------:|-----------------------------:|----------------:|
|system-notification-management-update.component.ts | 91% | ✅ |
|course-update.component.ts| 86% | ✅ |my changes have been covered|
|course.model.ts| 100%| ✅ |
|exercise.model.ts| 90%| ✅ |
|exam-update.component|97% |✅|
|file-upload-exercise-update.component.ts | 88% | ✅|
| modeling-exercise-update.component.ts | 89% | ✅| 
| quiz-exercise-detail.component.ts| 94% | ✅| 
| exercise.service.ts| 91% | ✅| 
| text-exercise-update.component.ts| 90% | ✅| 
| lecture-update.component.ts| 94% | ✅|
| lecture-update-wizard.component.ts| 95% | ✅|
| date-time-picker.component.ts| 75% | ✅ |my changes have been covered|
